### PR TITLE
Multichannel Implementation

### LINKED ISSUE #34: Handle Multi-Channel Audio Files Better
A current problem I am dealing with is multi-channel audio files.  The current method for handling this with fastai_audio is to compress them all into one channel, but I think this could be much more valuable to keep separately especially when multiple people are having a conversation and are being recorded on different channels.  How should fastai handle this without compressing these down into one channel?  

My thought is that each channel would be its own tensor so if there is a single channel, the shape would be (1,# of frames) when you added the second channel in, it would be (2,# of frames).  This would allow us to easily keep everything on the same timeline and make it generalizable to N channels.  How many places would this need to be handled?  I will start working through this if it is something that has interest.  My goal is to not have different code for 1-channel vs N-channel processors so I would want it to handle that the same way if possible.  Interested in getting feedback from other people that are familiar with the library.

### PR DESCRIPTION
**Pull request by kevinbird15**
Closes #34 

This pull request will allow for multi-channel audio files to be used without being compressed into a single channel.  I have ran through the 02_Features notebook and equally good results to what I was getting before adding multichannel support. The notebook `Multi-Channel Audio Files` in experiments is where most of the testing was completed for the new multi-channel functionality although a few tests were also added to the pytests.  There are warnings when the pytest tests are ran, but these are due to external libraries.  

To test this, open experiments/Multi-Channel Audio Files.ipynb and run through the notebook.

### CODE REVIEW

File: audio/audio.py [hunk 1]
```diff
@@ -16,38 +16,26 @@ class AudioItem(ItemBase):
...
-            sig, sr = torchaudio.load(item)
-            return AudioItem(sig, sr, path=Path(item))
-        if isinstance(item, (tuple, np.ndarray)):
-            return AudioItem(item)
 
     def show(self, title: [str] = None, **kwargs):
         print(f"File: {self.path}")
         print(f"Total Length: {round(self.duration, 2)} seconds")
         self.hear(title=title)
+        print(f"Number of Channels: {len(self.get_spec_images())}")
```
Comment (rbracco) [inline audio/audio.py]: It would be nice if it displayed each channel to keep things from being confusing when there are multiple channels and they use the delta. Currently it just displays 6 images with no labels (order is channel 1 sg, channel 1 delta, channel 1 accelerate, channel 2 sg, channel 2 delta, channel 2 accelerate). Ideal would probably be Channel 1: (3 images) Channel 2: (3 images)

File: audio/audio.py [hunk 2]
```diff
@@ -16,38 +16,26 @@ class AudioItem(ItemBase):
...
         for im in self.get_spec_images(): 
             display(im)
             print(f"Shape: {im.shape[1]}x{im.shape[2]}")
```
Comment (rbracco) [inline audio/audio.py]: Maybe use im.shape[-2] and im.shape[-1] for consistency with the rest of the code

File: audio/learner.py [hunk 3]
```diff
@@ -47,14 +47,14 @@ def audio_learner(data:AudioDataBunch, base_arch:Callable=models.resnet18, metri
     learn = cnn_learner(data, base_arch, cut=cut, metrics=metrics, pretrained=pretrained, lin_ftrs=lin_ftrs, ps=ps,
                         custom_head=custom_head, split_on=split_on, bn_final=bn_final, init=init,
                         concat_pool=concat_pool, **kwargs)
-    channels = _calc_channels(data.config)
+    channels = _calc_channels(data)
```
Comment (rbracco) [inline audio/learner.py:50]: Maybe use this PR as an opportunity to change the default pretrained setting to True for the audio_learner on line 41.

Comment (kevinbird15) [reply audio/learner.py:50]: I think we should do that in a separate PR since it isn't required for Multichannel.

**Comment by mogwai** [comment]
Are all of the audio files that you've included in this PR being used?

**Comment by kevinbird15** [comment]
I believe so.  The only one I'm not confident we need is openairmarket03.wav since we automatically download it.

File: audio/audio.py (hunk 2 above)
Comment (kevinbird15) [reply audio/audio.py]: Agreed, I updated this

File: audio/audio.py (hunk 1 above)
Comment (kevinbird15) [reply audio/audio.py]: Let me try to work through this.  I understand what you're saying, but have to figure out how these labels will be generated

File: audio/data.py [hunk 4]
```diff
@@ -258,66 +262,86 @@ def __init__(self, items, path, config=AudioConfig(), **kwargs):
...
             config.cache_dir = path / cd
         self.config = config
+        self.nchannels = 1 if self.config.downmix else None
         self.copy_new += ['config']
 
-    def open(self, item) -> AudioItem:
-        p = Path(item)
-        if self.path is not None and not p.exists() and str(self.path) not in str(item): p = self.path/item
-        if not p.exists(): raise FileNotFoundError(f"Neither '{item}' nor '{p}' could be found")
-        if not str(p).lower().endswith(AUDIO_EXTENSIONS): raise Exception("Invalid audio file")
```
Comment (mogwai) [inline audio/data.py:267]: This code was written to make it easier to work with from_csv and from_df. Why did you remove it?

File: audio/data.py [hunk 5]
```diff
@@ -355,3 +379,9 @@ def from_folder(cls, path: PathOrStr = '.', extensions: Collection[str] = None,
         if not extensions:
             extensions = AUDIO_EXTENSIONS
         return cls(get_files(path, extensions, recurse), path, **kwargs)
+
+def open_audio(fn:PathOrStr, after_open:Callable=None)->AudioItem:
+    if not str(fn).lower().endswith(AUDIO_EXTENSIONS): raise Exception("Invalid audio file")
+    sig, sr = torchaudio.load(fn)
+    if after_open: x = after_open(sig, sr)
```
Comment (mogwai) [inline audio/data.py]: After open is an Interesting feature but is it really needed?

File: audio/data.py [hunk 6]
```diff
@@ -258,66 +262,86 @@ def __init__(self, items, path, config=AudioConfig(), **kwargs):
...
-                if cfg.duration and cfg._processed:
-                    mel, start, end = tfm_crop_time(mel, cfg._sr, cfg.duration, cfg.sg_cfg.hop, cfg.pad_mode)
-                return AudioItem(spectro=mel, path=item, max_to_pad=cfg.max_to_pad, start=start, end=end)
-
-        sig, sr = torchaudio.load(str(p))
-        if(cfg._sr is not None and sr != cfg._sr):
-            raise ValueError(f'''Multiple sample rates detected. Sample rate {sr} of file {str(p)} 
-                                does not match config sample rate {cfg._sr} 
+    def open(self, fn): # file name, it seems
+        item = open_audio(fn)
```
Comment (mogwai) [inline audio/data.py]: You've lost the lazy loading here. 

Previously we would only open the audio file if the cache wasn't found and the use_spectro was true. The code was a bit complex but it was faster because of this. Could you add that back in?

File: audio/data.py [hunk 7]
```diff
@@ -258,66 +262,86 @@ def __init__(self, items, path, config=AudioConfig(), **kwargs):
...
+            if self.config.downmix:
+                item.sig = DownmixMono(channels_first=True)(item.sig)
+
+            elif(self.nchannels != item.nchannels):
+                raise ValueError(f'''Multiple channel sizes detected. Channel size {item.nchannels} of file 
+                                {fn} does not match others' channel size of {self.nchannels}. A dataset may
+                                not contain different number of channels. Please set downmix=true in AudioConfig or 
+                                separate files with different number of channels.''')
+
+    def add_spectro(self, item: AudioItem):
```
Comment (mogwai) [inline audio/data.py]: Really like the break down of all the logic into methods. Was something that was clearly needed :)

File: audio/data.py [hunk 8]
```diff
@@ -258,66 +262,86 @@ def __init__(self, items, path, config=AudioConfig(), **kwargs):
...
 
     def get(self, i):
-        item = self.items[i]
-        if isinstance(item, AudioItem): return item
-        if isinstance(item, (str, PosixPath, Path)):
-            if not ('/') in str(item): return self.open(self.path/item)
-            else:                      return self.open(item)
-        raise TypeError(f"Can't handle type {type(item)}, only AudioItem, str, Path or PosixPath")  
+        fn = super().get(i)
+        res = self.open(fn)
```
Comment (mogwai) [inline audio/data.py]: Why not just `return self.open(fn)`

File: tests/test_audio_data.py [hunk 9]
```diff
@@ -67,15 +67,39 @@ def test_cache_segment(random_item):
...
 
 def test_can_open():
-    p = Path('data/Right_whale.wav')
-    AudioItem.open(p)
-
+    p = Path('data/misc/whale/Right_whale.wav')
+    output = open_audio(p)
+    assert isinstance(output, AudioItem)
 
 def test_can_from_folder():
```
Comment (mogwai) [inline tests/test_audio_data.py:75]: Add a test for from_csv if you could as well

**Review by mogwai — CHANGES_REQUESTED** [review]

File: audio/audio.py (hunk 1 above)
Comment (kevinbird15) [reply audio/audio.py]: I don't know how to handle delta here.  Everything else I can do, but we don't know if delta is true here.  Unless I Just have an argument to show called is_delta: bool.  Then the user would have to add is_delta=True if they wanted to see the show with Delta included

File: audio/data.py (hunk 8 above)
Comment (kevinbird15) [reply audio/data.py]: So you want get to look like this: 

```
    def get(self, i):
        fn = super().get(i)
        return self.open(fn)
```

Comment (mogwai) [reply audio/data.py]: Only a small thing but saves one line :)

File: audio/data.py (hunk 5 above)
Comment (rbracco) [reply audio/data.py]: I like it and think we'll need the pattern anyway, it's opening the door to callbacks right?

File: audio/data.py (hunk 8 above)
Comment (kevinbird15) [reply audio/data.py]: Just making sure I was understanding what you were saying.  I'll change that.

File: audio/data.py (hunk 6 above)
Comment (rbracco) [reply audio/data.py]: Can confirm. When we are training on spectrograms there's no reason to load the audio as it takes time and we don't use it. If the spectrogram is coming from the cache, we only want to load the audio when the user requests to see it via `.show()`.  The easiest fix that preserves the new pipeline and improved modularity would be to...

In `open_audio()` check first to see if they are using spectros, and if the spectro is already cached. If so, grab it and return a new `AudioItem` with that spectro and sig/sr = None. If there isn't a spectro, load the audio file and return an `AudioItem` containing sig/sr but spectro=None (or if `use_spectro` = True you could call create_spectro and generate it right there). 

Doing this means that you'll have to go through the rest of the pipeline and see what is affected and fix it. But it seems like the best way to preserve the structure you've set up by generating an `AudioItem` up front and handing it off to each function for validation/spectro generation...etc

File: audio/audio.py [hunk 10]
```diff
@@ -76,36 +64,12 @@ def apply_tfms(self, tfms):
     @property
     def shape(self): return self.data.shape
 
-    def _reload_signal(self):
```
Comment (rbracco) [inline audio/audio.py:79]: All of this code had a purpose, although there may be a better way to do it that I'm not aware of.

`_reload_signal` is here so that when the user calls `.show_batch()` or views `AudioItem`s in some way that shows the signal and spectrogram, the signal will be reloaded. Otherwise we have to load it every time we grab an item for training even though the audio isn't being used.

The internal variables `_sig` and `_sr` were being used to assist with that lazy loading.

File: audio/audio.py (hunk 1 above)
Comment (kevinbird15) [reply audio/audio.py]: image

Comment (kevinbird15) [reply audio/audio.py]: Is this acceptable for the displaying the channel?

File: audio/data.py (hunk 4 above)
Comment (kevinbird15) [reply audio/data.py:267]: we were trying to simplify the code and I believe the checks were lost when we split things up.  I'm going to move where these are happening some, but I believe I've added everything back in.

File: audio/data.py (hunk 6 above)
Comment (kevinbird15) [reply audio/data.py]: This should be working now.  I believe it is lazy, but would like you to recheck it now.

File: audio/data.py (hunk 5 above)
Comment (kevinbird15) [reply audio/data.py]: I added something that utilizes after_open.  I think it is a useful feature that should be kept in.  Since config isn't passed into open_audio, it gives the ability to do things that would be much more complex without it.

File: audio/data.py [hunk 11]
```diff
@@ -258,66 +262,86 @@ def __init__(self, items, path, config=AudioConfig(), **kwargs):
...
-            start, end = None, None
-            if cfg.duration and cfg._processed: 
-                mel, start, end = tfm_crop_time(mel, cfg._sr, cfg.duration, cfg.sg_cfg.hop, cfg.pad_mode)
-        return AudioItem(sig=sig.squeeze(), sr=sr, spectro=mel, path=item, start=start, end=end)
+
+        if self.nchannels is None: self.nchannels = item.nchannels
+            
+        if item.nchannels > 1:
+            if self.config.downmix:
+                item.sig = DownmixMono(channels_first=True)(item.sig)
```
Comment (kevinbird15) [inline audio/data.py]: I think this may need to happen somewhere else.  When it happens here, I don't believe it will every play as monochannel.  The Spectrogram is one channel, but the actual audio when doing `.show()` is multichannel still.  May be fixed after processing further?  Need to dig in more here.

File: audio/audio.py (hunk 10 above)
Comment (kevinbird15) [reply audio/audio.py:79]: Ok, that definitely makes sense now that I've tried reimplementing lazy loading.  I had almost an identical solution so I tried changing the names back and only using `_sig` and `_sr` to get `sig` and `sr` populated

File: tests/test_audio_data.py (hunk 9 above)
Comment (kevinbird15) [reply tests/test_audio_data.py:75]: This is still open. Didn't have a chance to add it tonight although I believe it will work now

File: audio/data.py (hunk 11 above)
Comment (kevinbird15) [reply audio/data.py]: I moved this to happen with the other preprocessing.

File: tests/test_audio_data.py (hunk 9 above)
Comment (rbracco) [reply tests/test_audio_data.py:75]: I have several personal projects outside the repo that are dependent on from_csv. I'll see if they work with this version and once we are aligned with fastai_V2 we can add proper testing with pytest.

File: audio/audio.py (hunk 1 above)
Comment (rbracco) [reply audio/audio.py]: This looks pretty good but maybe we should move to only displaying the second part if there are mult images per channel. Most of the time there wont be mult images so having channel 0.0, channel 1.0...etc is maybe not ideal. Leave it for now though, it looks great in the multichannel delta case.

File: audio/data.py [hunk 12]
```diff
@@ -260,64 +288,75 @@ def __init__(self, items, path, config=AudioConfig(), **kwargs):
...
-                os.makedirs(image_path.parent, exist_ok=True)
-                torch.save(mel, image_path)
-                _record_cache_contents(cfg, [image_path])
-            start, end = None, None
-            if cfg.duration and cfg._processed: 
-                mel, start, end = tfm_crop_time(mel, cfg._sr, cfg.duration, cfg.sg_cfg.hop, cfg.pad_mode)
-        return AudioItem(sig=sig.squeeze(), sr=sr, spectro=mel, path=item, start=start, end=end)
+        if(self.config._nchannels != item.nchannels):
+            raise ValueError(f'''Multiple channel sizes detected. Channel size {item.nchannels} of file 
+                                {fn} does not match others' channel size of {self.config._nchannels}. A dataset may
```
Comment (rbracco) [inline audio/data.py]: `fn` should be `item.path` here

File: tests/test_audio_data.py (hunk 9 above)
Comment (mogwai) [reply tests/test_audio_data.py:75]: Great let us know when you've confirmed

File: audio/data.py [hunk 13]
```diff
@@ -214,12 +232,19 @@ def _pre_process(self):
         
         if len(x.items) > 0:
             if not cfg.resample_to: _set_sr(x.items[0], x.config, x.path)
-            if cfg.remove_silence or cfg.segment_size or cfg.resample_to:
+            _set_nchannels(x.items[0], x.config)
```
Comment (rbracco) [inline audio/data.py]: We only need to call this function once, so wrap it in a statement that checks if config._nchannels is already set...
`if cfg._nchannels is None: _set_nchannels(x.items[0], x.config)`

File: audio/data.py [hunk 14]
```diff
@@ -214,12 +232,19 @@ def _pre_process(self):
         
         if len(x.items) > 0:
             if not cfg.resample_to: _set_sr(x.items[0], x.config, x.path)
```
Comment (rbracco) [inline audio/data.py]: This should technically be 
`if cfg.resample_to is None` but probably doesnt matter since we'll  never have 0 sample rate

File: audio/data.py [hunk 15]
```diff
@@ -260,64 +288,75 @@ def __init__(self, items, path, config=AudioConfig(), **kwargs):
...
-            if cfg.delta: mel = torch.stack([m.squeeze(0) for m in [mel, torchdelta(mel), torchdelta(mel, order=2)]]) 
-            if cfg.cache:
-                os.makedirs(image_path.parent, exist_ok=True)
-                torch.save(mel, image_path)
-                _record_cache_contents(cfg, [image_path])
-            start, end = None, None
-            if cfg.duration and cfg._processed: 
-                mel, start, end = tfm_crop_time(mel, cfg._sr, cfg.duration, cfg.sg_cfg.hop, cfg.pad_mode)
-        return AudioItem(sig=sig.squeeze(), sr=sr, spectro=mel, path=item, start=start, end=end)
+        if(self.config._nchannels != item.nchannels):
```
Comment (rbracco) [inline audio/data.py]: Need to have a check here that `self.config._nchannels` isn't None.

We want users to be able to generate AudioLists with many sample rates or channels, it's only at train time that this is a problem. Preprocessing to ensure single SR and n_channels is done at the same time as labeling, that's why there are lines to set `config._sr` and `config._nchannels` to be their correct values. Before that point they are both `None` and if they are `None` that means we don't want to validate.

**Comment by kevinbird15** [comment]
I think everything is good now as far as I can tell.
